### PR TITLE
fix: data race collecting GPU slow/frozen frames and framerates

### PR DESCRIPTION
**Pull request by armcknight**
## :scroll: Description

Adds some extra guards in the case where we don't have all the data we expect, which is when we do have frozen/slow frame info but seem to not have any frame rate info. This avoids some unnecessary work and of course the attempt to insert nil into a dictionary literal, which is causing a crash.

Also fixes a small bug where we clear the profiling GPU frame rate data structure when we stop the profiler due to timing out or going to the background; in these cases, we don't want to throw those away until after the associated span has finished and we serialize the profiler data for it. So we only clear the data when stopping normally, which is whenever the last in-flight span has finished in SentryTracer.

## :bulb: Motivation and Context

To fix https://github.com/getsentry/sentry-cocoa/issues/3082

## :green_heart: How did you test it?

I was unable to reproduce. One theory was that if was a data race between a timed out profiler (since the customer's stack trace shows the span actually timed out as well) and another subsequent profiler, but I don't see a data race because reads and writes to SentryFramesTracker's profiling timestamp data structures are all synchronized with lock_guard.

## :pencil: Checklist

You have to check all boxes before merging:

- [ ] I reviewed the submitted code.
- [ ] I added tests to verify the changes.
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [ ] No breaking change or entry added to the changelog.
- [ ] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

**Review by indragiek — CHANGES_REQUESTED** [review]

File: Sources/Sentry/SentryProfiler.mm [hunk 1]
```diff
@@ -118,8 +118,13 @@
 sliceGPUData(SentryFrameInfoTimeSeries *frameInfo, SentryTransaction *transaction,
     BOOL useMostRecentRecording)
 {
+    __block NSNumber *nearestPredecessorValue = frameInfo.firstObject[@"value"];
+    if (UNLIKELY(useMostRecentRecording && nearestPredecessorValue == nil)) {
```
[the remark this answers is not in the document]
Comment (armcknight) [reply Sources/Sentry/SentryProfiler.mm:122]: Yes we need to check for it. nearestPredecessorValue is only relevant if useMostRecentRecording is true. This comes down to the decoupled nature of slow/frozen frame recordings vs frame rate recordings. We have to do the work of zipping those two disparate datasets together.

File: Sources/Sentry/SentryProfiler.mm [hunk 2]
```diff
@@ -183,6 +188,45 @@
     return result;
 }
 
+NSMutableDictionary<NSString *, id> *_Nullable serializeGPUMetrics(SentryTransaction *transaction)
+{
+    const auto gpuMetrics = [NSMutableDictionary<NSString *, id> dictionary];
+
+    if (UNLIKELY(_gCurrentFramesTracker.currentFrames.frameRateTimestamps.count == 0)) {
```
[the remark this answers is not in the document]
Comment (armcknight) [reply Sources/Sentry/SentryProfiler.mm:195]: Yes I mentioned this in the PR description but I have been unable to tease out any races if they actually exist. Please see the description for details.

File: Sources/Sentry/SentryProfiler.mm (hunk 1 above)
[the remark this answers is not in the document]
Comment (armcknight) [reply Sources/Sentry/SentryProfiler.mm:122]: It's called up to three times per profile, but agree that ofc it's not nearly as frequent as backtrace sampling. I figured two things: 1. every little bit helps and 2. self-documenting. Would you like it removed?

File: Sources/Sentry/SentryProfiler.mm [hunk 3]
```diff
@@ -118,8 +118,13 @@
 sliceGPUData(SentryFrameInfoTimeSeries *frameInfo, SentryTransaction *transaction,
     BOOL useMostRecentRecording)
 {
+    __block NSNumber *nearestPredecessorValue = frameInfo.firstObject[@"value"];
+    if (UNLIKELY(useMostRecentRecording && nearestPredecessorValue == nil)) {
+        // This is unlikely because in the only case when useMostRecentRecording is YES, for frame
+        // rates, we won't call this function unless
```
[the remark this answers is not in the document]
Comment (armcknight) [reply Sources/Sentry/SentryProfiler.mm:124]: argh lol. should read "unless there is at least one frame rate recorded."

**Comment by armcknight** [comment]
Just realized this still needs a fix. If we just automatically take the first frame rate recorded, it's theoretically possible it wasn't recorded until after the transaction (if there really is some kind of data race going on)... we need to not only nil check it at the top of sliceGPUData but also check that it's chronologically ordered appropriately.

**Comment by armcknight** [comment]
I fixed one more issue with that commit above, but I need to get the tests working again, the mocking needs to be changed slightly to account for a first frame rate of relative timestamp 0 for every profile. This is just a bit of a pain for the concurrent transaction test case, needs a little more work.

**Pull request closed without merging**